src/microbit/model/image.py

```diff
@@ -0,0 +1,290 @@
+from . import constants as CONSTANTS
+
+
+class Image:
+
+    # implementing image model as described here:
+    # https://microbit-micropython.readthedocs.io/en/latest/image.html
+
+    def __init__(self, *args, **kwargs):
+
+        # Depending on the number of arguments
+        # in constructor, it treat args differently.
+
+        if len(args) == 0:
+            # default constructor
+            self.__LED = self.__string_to_array(CONSTANTS.BLANK_5X5)
+        elif len(args) == 1:
+            pattern = args[0]
+            if isinstance(pattern, str):
+                self.__LED = self.__string_to_array(pattern)
+            else:
+                raise TypeError("Image(s) takes a string")
+        else:
+
+            width = args[0]
+            height = args[1]
+
+            if width < 0 or height < 0:
+                # This is not in original, but ideally,
+                # image should fail non-silently
+                raise ValueError(CONSTANTS.INDEX_ERR)
+
+            if len(args) == 3:
+                # This option is for potential third bytearray arguments
+                byte_arr = args[2]
+                self.__LED = self.__bytes_to_array(width, height, byte_arr)
+            else:
+                self.__LED = self.__create_leds(width, height)
+
+    def width(self):
+        if len(self.__LED) > 0:
+            return len(self.__LED[0])
+        else:
+            return 0
+
+    def height(self):
+        return len(self.__LED)
+
+    def set_pixel(self, x, y, value):
+        try:
+            if not self.__valid_pos(x, y):
+                raise ValueError(CONSTANTS.INDEX_ERR)
+            elif not self.__valid_brightness(value):
+                raise ValueError(CONSTANTS.BRIGHTNESS_ERR)
+            else:
+                self.__LED[y][x] = value
+        except TypeError:

```

Comment:
When do we expect this TypeError to be raised?
Is it in: `self.__LED[y][x] = value` ?

Reply:
Oops, that was from before when the internal data structure was a tuple for non-editable instances. I'll remove it.